### PR TITLE
[SCB-370] Improve latency and max precision

### PR DESCRIPTION
**Pull request by zhengyangyong**
…ency and max latency precision of metrics

Signed-off-by: zhengyangyong <yangyong.zheng@huawei.com>

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
If operation make two request with latency 1234567 nanoseconds and 7654321 nanoseconds
Before :
latency (average) output will be 4 milliseconds,and max output will be 7 milliseconds (value after the decimal point lost)
Now :
latency (average) output will be 4.444444 milliseconds,and max output will be 7.654321 milliseconds

### CODE REVIEW

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public Double getValue() {
+      return this.getValue(0);
+    }
+
+    @Override
+    public Double getValue(int pollerIndex) {
+      //we need direct div for keep value after the decimal point
+      long nanoValue = timer.getValue();
+      return (double) nanoValue / (double) 1000000;
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: Output is ms , not second?
this is not compatible with servo and spectator

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Override
+    public MonitorConfig getConfig() {
+      return timer.getConfig().withAdditionalTag(Tags.newTag(MetricsConst.TAG_STATISTIC, "latency"));
+    }
+  }
+
+  class MonitorForTimerMax implements Monitor<Double> {
+
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: max and total time already in basicTimer, why we need so many  logic ……

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java (hunk 1 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: servo default ouput is ms

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java (hunk 2 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: I will not direct call getTotal() in basicTimer,this method is private for BasicTimer 
Metrics output will get all register Monitors from MonitorRegistry  and call it's getValue()
more detail can see MonitorManager.measure()

Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: MonitorRegistry holds all registered Monitors (timer,counter,gauge etc...) , getRegistedMonitors() return base Monitor<T> interface collection;
BaiscTimer extends Monitor<Long>, getValue()  will return Long type value,so only one logic is div 1000000

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java (hunk 1 above)
Comment (wujimin) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: where is the code, what i see is second.

Comment (wujimin) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: it seems that you confused input unit and output unit

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java (hunk 2 above)
Comment (wujimin) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: basicTimer provide monitors, include: totalTimeFactor, count, min, max  
totalTimeFactoris FactorMonitor wrapped totalTime  
totalTime is StepCounter  
FactorMonitor .getValue delegate to StepCounter.getValue  
so output is second  

did i missed something?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java (hunk 1 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: public BasicTimer(MonitorConfig config) {
        this(config, TimeUnit.MILLISECONDS);
    }

BasicTimer(MonitorConfig config, TimeUnit unit, Clock clock) {
...
timeUnitNanosFactor = 1.0 / timeUnit.toNanos(1);
...
}

private double getTotal(int pollerIndex) {
        return totalTime.getCurrentCount(pollerIndex) * timeUnitNanosFactor;
    }

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java (hunk 2 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/HighPrecisionBasicTimer.java]: I direct use getValue() implemented in basicTimer:
    @Override
    public Long getValue(int pollerIndex) {
        final long cnt = count.getCurrentCount(pollerIndex);
        final long value = (long) (getTotal(pollerIndex) / cnt);
        return (cnt == 0) ? 0L : value;
    }

set it's output unit to nano then div 1000000

**Comment by zhengyangyong** [comment]
从内部MonitorRegistry获取值改为直接输出原始值不加工，之后用户使用MetricsUtils.convertMeasurements依据需求转换单位，现在已有场景：
1./metrics Publish 使用**秒**
2.对接Prometheus 使用毫秒
3.perf 打印使用毫秒
4.写文件的例子使用毫秒

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MetricsUtils.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+              convertFunctions.get(unit).apply(measurement.getValue()));
+        } else {
+          metrics.put(getMonitorKey(measurement.getKey()), measurement.getValue());
+        }
+      }
+      return metrics;
+    }
+    //no need support MINUTES,HOURS,DAYS because latency under this unit is unsuitable
+    throw new ServiceCombException(
+        "illegal unit : " + String.valueOf(unit) + ", only support NANOSECONDS,MICROSECONDS,MILLISECONDS and SECONDS");
```
Comment (yangbor) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MetricsUtils.java]: Metrics calculated in minutes or even hours is normal, we should support at least to hours.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MetricsUtils.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import org.apache.servicecomb.foundation.common.exceptions.ServiceCombException;
+import org.apache.servicecomb.foundation.metrics.MetricsConst;
+
+import com.netflix.servo.monitor.MonitorConfig;
+import com.netflix.servo.tag.Tag;
+import com.netflix.servo.tag.TagList;
+import com.netflix.servo.tag.Tags;
+
+public class MetricsUtils {
```
Comment (yangbor) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MetricsUtils.java]: XXXUtils suggests a bad design, later developers may put any unrelated things into this class.
Move the getMonitorKey to MonitorManager and rename this class to MeasurementConversion or something suitable.

**Pull request closed without merging**